src/translator/service.cpp

```diff
@@ -6,33 +6,51 @@
 #include <utility>
 
 inline std::vector<marian::Ptr<const marian::Vocab>>
-loadVocabularies(marian::Ptr<marian::Options> options) {
+loadVocabularies(marian::Ptr<marian::Options> options,
+                 std::vector<marian::bergamot::AlignedMemory>&& vocabMemories,
+                 std::vector<size_t>&& vocabIndices) {
   // @TODO: parallelize vocab loading for faster startup
-  auto vfiles = options->get<std::vector<std::string>>("vocabs");
-  // with the current setup, we need at least two vocabs: src and trg
-  ABORT_IF(vfiles.size() < 2, "Insufficient number of vocabularies.");
-  std::vector<marian::Ptr<marian::Vocab const>> vocabs(vfiles.size());
-  std::unordered_map<std::string, marian::Ptr<marian::Vocab>> vmap;
-  for (size_t i = 0; i < vocabs.size(); ++i) {
-    auto m =
-        vmap.emplace(std::make_pair(vfiles[i], marian::Ptr<marian::Vocab>()));
-    if (m.second) { // new: load the vocab
-      m.first->second = marian::New<marian::Vocab>(options, i);
-      m.first->second->load(vfiles[i]);
+  std::vector<marian::Ptr<marian::Vocab const>> vocabs;
+  if(!vocabMemories.empty()){
+    // load vocabs from buffer
+    ABORT_IF(vocabIndices.size() < 2, "Insufficient number of vocabularies.");
+    vocabs.resize(vocabMemories.size());
+    for (size_t i = 0; i < vocabs.size(); i++) {
+      marian::Ptr<marian::Vocab> vocab = marian::New<marian::Vocab>(options, i);
+      vocab->loadFromSerialized(absl::string_view(vocabMemories[vocabIndices[i]].begin(),
+                                                  vocabMemories[vocabIndices[i]].size()));
+      vocabs[i] = vocab;
+    }
+  } else {
+    // load vocabs from file
+    auto vfiles = options->get<std::vector<std::string>>("vocabs");
+    // with the current setup, we need at least two vocabs: src and trg
+    ABORT_IF(vfiles.size() < 2, "Insufficient number of vocabularies.");
+    vocabs.resize(vfiles.size());
+    std::unordered_map<std::string, marian::Ptr<marian::Vocab>> vmap;
+    for (size_t i = 0; i < vocabs.size(); ++i) {
+      auto m = vmap.emplace(std::make_pair(vfiles[i], marian::Ptr<marian::Vocab>()));
+      if (m.second) { // new: load the vocab
+        m.first->second = marian::New<marian::Vocab>(options, i);
+        m.first->second->load(vfiles[i]);
+      }
+      vocabs[i] = m.first->second;
     }
-    vocabs[i] = m.first->second;
   }
   return vocabs;
 }
 
 namespace marian {
 namespace bergamot {
 
-Service::Service(Ptr<Options> options, AlignedMemory modelMemory, AlignedMemory shortlistMemory) 
-    : requestId_(0), options_(options), vocabs_(std::move(loadVocabularies(options))),
+Service::Service(Ptr<Options> options, AlignedMemory modelMemory, AlignedMemory shortlistMemory,
+                 std::vector<AlignedMemory> vocabMemories, std::vector<size_t> vocabIndices)
```

Comment:
@qianqianzhu We have 3, maybe 4 arguments related to bytearray now, one more (shortlist) incoming. Now might be a good time to:

* https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#i23-keep-the-number-of-function-arguments-low

Reply:
Do you want to wrap all AlignedMemory objects together?

Reply:
I'm thinking a self describing structure like the following (I gave quick names, you can probably come up with something easier) which should make code in a lot of places neater. 

```
struct VocabBundle {
	std::vector<AlignedMemory> uniqueVocabs;
	set::vector<size_t> vocabIndices;
};

struct MemoryBundle {
	AlignedMemory model;
	AlignedMemory shortlist;
	VocabBundle vocabs;
	AlignedMemory ssplitPrefixFile;
};
```
Of course you don't have `ssplitPrefixFile` yet, but you can see how further changes become simplified with this structure. There's also discussions surrounding if MemoryBundle (or a subset) will be the single binary Mozilla provides us.

Reply:
Would be nice to have, but I think it's outside of the scope of this pr. Refactoring may come later.
Technically, we could have arbitrary number of models, shortlists and vocabularies loaded too (marian supports that), but we don't support that.

Reply:
I'm ok with doing the struct instead of multiple args in a different PR.

Reply:
I will leave this refactoring later (in a new PR).